### PR TITLE
WIP: Generic IDL introspection

### PR DESCRIPTION
**Pull request by bombela**
This adds an API for generic IDL introspection. Here, an idl is just a file with an optional list of idl includes. A set of method is provided to conveniently build a tree of IDLs from IDLs with includes.

Another PR will integrate thrift IDL to generic IDL conversion.

### CODE REVIEW

File: internal/introspection/handler.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package introspection
+
+type IntrospectableHandler interface {
```
Comment (peter-edge) [inline internal/introspection/handler.go]: Every public type needs a comment, I'm actually surprised this passes linting, does it?

File: internal/introspection/idlmodule.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+package introspection
+
+import "strings"
+
+// IDLModule is a generic IDL module. For example, a thrift file or a protobuf
+// one.
+type IDLModule struct {
+	FilePath   string      `json:"filePath"`
+	SHA1       string      `json:"sha1"`
```
Comment (peter-edge) [inline internal/introspection/idlmodule.go]: Maybe `Checksum`, to not lock into SHA1? You could also have `ChecksumType`.

File: internal/introspection/idlmodule.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	Includes   []IDLModule `json:"includes,omitempty"`
+	RawContent string      `json:"rawContent,omitempty"`
```
Comment (peter-edge) [inline internal/introspection/idlmodule.go]: Content? Also may want to explain a bit more here what is included (comments on fields?)

File: internal/introspection/idlmodule.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+// Iteration is aborted if cb returns true.
+func (im *IDLModule) Flatmap(cb func(*IDLModule) bool) {
+	ft := newIDLFlattener(cb)
+	ft.Collect(im)
+}
+
+// IDLModules is a sortable slice of IDLModule.
+type IDLModules []IDLModule
+
+func (ims IDLModules) Len() int {
```
Comment (peter-edge) [inline internal/introspection/idlmodule.go]: I won't comment on every one of these, but as I said, in general you need a godoc comment on every public type.

File: internal/introspection/idlmodule.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	Includes   []IDLModule `json:"includes,omitempty"`
+	RawContent string      `json:"rawContent,omitempty"`
+}
+
+// Flatmap iterates recursively over the idl module and its includes. The cb
+// function is called onces for every unique idl module (based on FilePath).
+// Iteration is aborted if cb returns true.
+func (im *IDLModule) Flatmap(cb func(*IDLModule) bool) {
+	ft := newIDLFlattener(cb)
+	ft.Collect(im)
```
Comment (peter-edge) [inline internal/introspection/idlmodule.go]: nit: why not just one-line it, `newIDLFlattener(cb).Collect(im)`

File: internal/introspection/idlmodule.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package introspection
+
+import "strings"
+
+// IDLModule is a generic IDL module. For example, a thrift file or a protobuf
```
Comment (peter-edge) [inline internal/introspection/idlmodule.go]: Module is slightly overloaded here - in protobuf, you have explicit packages named at the top of the file that may not correspond to the file name.

File: internal/introspection/handler.go (hunk 1 above)
Comment (bombela) [inline internal/introspection/handler.go]: it did not pass linting! I literally didnt run the linter before.

File: internal/introspection/idlmodule.go (hunk 6 above)
Comment (bombela) [inline internal/introspection/idlmodule.go]: `IDLFile` perhaps? let me know what you think.

File: internal/introspection/idlmodule.go (hunk 4 above)
Comment (bombela) [inline internal/introspection/idlmodule.go]: linter is happy with uncommented standard sort methods for what i can tell.

File: api/middleware/inbound.go [hunk 7]
```diff
@@ -71,6 +75,13 @@ func (h unaryHandlerWithMiddleware) Handle(ctx context.Context, req *transport.R
 	return h.i.Handle(ctx, req, resw, h.h)
 }
 
+func (h unaryHandlerWithMiddleware) Introspect() *introspection.Handler {
```
Comment (willhug) [inline api/middleware/inbound.go]: seems unrelated...

File: internal/introspection/idlfile.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	for i := range ims {
+		ft.Collect(&ims[i])
+	}
+}
+
+// IDLTree is a tree of IDLs. A tree can have directories. Dir maps directory
+// names to IDLtrees. And of course, a tree can have any number of idl modules.
+type IDLTree struct {
+	Dir     map[string]*IDLTree `json:"dir,omitempty"`
+	Modules IDLFiles            `json:"modules,omitempty"`
```
Comment (willhug) [inline internal/introspection/idlfile.go]: files

**Comment by willhug** [comment]
Talked offline, this is more of a WIP diff to determine the whether the structure proposed makes sense.  Before this can be landed/properly reviewed it needs to have tests.

File: internal/introspection/handler.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package introspection
+
+// IntrospectableHandler is what you think it is.
```
Comment (willhug) [inline internal/introspection/handler.go]: Is this for HandlerSpecs?  Is this different from a procedure?

**Comment by bombela** [comment]
Yes! thank you Will. I should have written a disclaimer atop my PR, this is duly noted. For this first review, let me know what you think of the overall API and style. Then I will add tests with any modification commanded by the review(s).

File: internal/introspection/idlfile.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// Compact replaces any tree with a single directory and no IDLFiles by it's
+// single child. The the lone directory name is appended to the parent's
+// directory. I would write an example, but go doc would likely fuck up
+// rendering it anyway.
```
Comment (willhug) [inline internal/introspection/idlfile.go]: please add a comment example here, it will be muuuuuuch easier to visualize

File: internal/introspection/idlfile.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+func (it *IDLTree) Compact() {
+	for _, l1tree := range it.Dir {
+		l1tree.Compact()
+	}
+	for l1dir, l1tree := range it.Dir {
+		if len(l1tree.Dir) == 1 && len(l1tree.Modules) == 0 {
```
Comment (willhug) [inline internal/introspection/idlfile.go]: we should invert this logic to reduce indentation.

```
if len(tree.Dir) != 1 || len(tree.Modules) != 0 {
  continue
}
...
```

File: internal/introspection/idlfile.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+	ft := newIDLFlattener(cb)
+	for i := range ims {
+		ft.Collect(&ims[i])
+	}
+}
+
+// IDLTree is a tree of IDLs. A tree can have directories. Dir maps directory
+// names to IDLtrees. And of course, a tree can have any number of idl modules.
+type IDLTree struct {
+	Dir     map[string]*IDLTree `json:"dir,omitempty"`
```
Comment (willhug) [inline internal/introspection/idlfile.go]: SubDirs?

File: internal/introspection/idlfile.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+				compactedDir := l1dir + "/" + l2dir
+				it.Dir = map[string]*IDLTree{compactedDir: l2tree}
+				break
+			}
+		}
+	}
+}
+
+type idlFlattener struct {
+	seen map[string]struct{}
```
Comment (willhug) [inline internal/introspection/idlfile.go]: seenDirs?

File: internal/introspection/idlfile.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+	cb   func(*IDLFile) bool
```
Comment (willhug) [inline internal/introspection/idlfile.go]: is cb for callback?

File: internal/introspection/idlfile.go [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+	Includes []IDLFile `json:"includes,omitempty"`
+
+	// Content is the IDL file content.
+	Content string `json:"content,omitempty"`
+}
+
+// Flatmap iterates recursively over the idl module and its includes. The cb
+// function is called onces for every unique idl module (based on FilePath).
+// Iteration is aborted if cb returns true.
+func (im *IDLFile) Flatmap(cb func(*IDLFile) bool) {
```
Comment (willhug) [inline internal/introspection/idlfile.go]: What kind of `cb` function would I pass in here? Why would I want to do it?

File: internal/introspection/idlfile.go [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+func (ib *idlTreeBuilder) Collect(m *IDLFile) {
+	if _, ok := ib.seen[m.FilePath]; ok {
+		return
+	}
+
+	ib.seen[m.FilePath] = struct{}{}
+	n := &ib.Root
+	parts := strings.Split(m.FilePath, "/")
+	for i, part := range parts {
+		if i == len(parts)-1 {
```
Comment (willhug) [inline internal/introspection/idlfile.go]: why don't we just strip the last part of the file path before iterating?

File: internal/introspection/idlfile.go [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+func (ib *idlTreeBuilder) Collect(m *IDLFile) {
+	if _, ok := ib.seen[m.FilePath]; ok {
+		return
+	}
+
+	ib.seen[m.FilePath] = struct{}{}
+	n := &ib.Root
+	parts := strings.Split(m.FilePath, "/")
```
Comment (willhug) [inline internal/introspection/idlfile.go]: directories

**Comment by willhug** [comment]
The style in this diff seems alright (nodes+files+trees probably aren't really contentious).  Not sure what the purpose of some of the attributes are without an overall goal for this project.  Might make sense to have a task.

File: api/middleware/inbound.go (hunk 7 above)
Comment (bombela) [inline api/middleware/inbound.go]: It actually is. Because generic idl introspection must be forwarded from handlers, trough middlewares, up to the table router.

File: internal/introspection/idlfile.go (hunk 15 above)
Comment (bombela) [inline internal/introspection/idlfile.go]: IDLFile represents an IDL file and its includes. Recursively, IDLFile includes are IDLFile's themselves. IDLFiles can also indirectly includes identical files. So an IDLFile is a DAG. The flatmap function iterates recursively trough this DAG, calling the `cb` function only once per IDLFile. If A includes B and C, and both B and C include D, then flatmap will call `cb` once for A, B, D and C. (there is no ordering guarantee).

Maybe I should have de-duplication as an additional step instead of backed into the flatmap operation. Let me know if this is what confused you.

File: internal/introspection/idlfile.go (hunk 13 above)
Comment (bombela) [inline internal/introspection/idlfile.go]: idlFlattener, the subject is an IDL. thus its about an idl file. I will rename into `seenIDLs`.

File: internal/introspection/idlfile.go (hunk 17 above)
Comment (bombela) [inline internal/introspection/idlfile.go]: the last element might not necessarily be a directory. the URI RFC refers to the this as `path segments`. I will rename to `segments`.

File: internal/introspection/handler.go (hunk 9 above)
Comment (bombela) [inline internal/introspection/handler.go]: yes, a handler spec is one building block of a procedure.

File: internal/introspection/idlfile.go (hunk 12 above)
Comment (bombela) [inline internal/introspection/idlfile.go]: Its the directory of the IDLTree. I believe singular is correct here.

File: internal/introspection/idlmodule.go (hunk 6 above)
Comment (peter-edge) [inline internal/introspection/idlmodule.go]: That might work

File: internal/introspection/idlfile.go [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+// IDLFile is a generic IDL file. For example, a thrift or protobuf IDL.
+type IDLFile struct {
+	// FilePath starts from an arbitrary root containing the file and all its
+	// includes.
+	FilePath string `json:"filePath"`
+
+	// Checksum of Content. The format is: "${type}:${hex}". For example, a
+	// sha1 checksum would look like:
+	// "sha1:d55f7ce6138889e9e45b62df3741820777c65fcd".
+	Checksum string `json:"checksum"`
```
Comment (peter-edge) [inline internal/introspection/idlfile.go]: I'm not the biggest fan of having a semantic representation in a string like this, what about `ChecksumType string` and `Checksum`, or `type ChecksumType int`?

Comment (bombela) [inline internal/introspection/idlfile.go]: I hear you.  Thing is, I would prefer to treat the checksum as simply as possible. I could have used a slice of bytes. But I have decided to go with the convention of a printable string. With this simple string, the json output is simple. I really do not thing there is any need to complicate this here. 

At first i had named the checksum SHA1. Which basically made yarpc forcing you to implement a specific hash type. In exchange consumer of yarpc know they have a single checksum/hash to deal with. With an enum, yarpc would now force you to deal with different checksum types. Worst, you can only emit checksum types vetted by yarpc now. With a hash type as a string (here, part of the string because why not), yarpc is removed from any decision on the matter.

So, as I see it, we have 3 choices:
 1. we pick a single checksum types, and stick with it.
 2. we offer a fixed list of checksum types.
 3. the checksum is a self-contained string.

I really want to start simple, and I am in favor of option 3.

Comment (peter-edge) [inline internal/introspection/idlfile.go]: To answer why not: I expect a struct field named `Checksum` to actually be a checksum. This is basically embedding an object with a custom wire format to this field. Why not just have `ChecksumType string` at least?
